modules/memberships/class-jetpack-memberships.php

```diff
@@ -0,0 +1,283 @@
+	 * Actual hooks initializing on init.
+	 */
+	public function init_hook_action() {
+		add_filter( 'rest_api_allowed_post_types', array( $this, 'allow_rest_api_types' ) );
+		add_filter( 'jetpack_sync_post_meta_whitelist', array( $this, 'allow_sync_post_meta' ) );
+		$this->setup_cpts();
+	}
+
+	/**
+	 * Sets up the custom post types for the module.
+	 */
+	private function setup_cpts() {
+		/*
+		 * PLAN data structure.
+		 */
+		$capabilities = array(
+			'edit_post'          => 'edit_posts',
+			'read_post'          => 'read_private_posts',
+			'delete_post'        => 'delete_posts',
+			'edit_posts'         => 'edit_posts',
+			'edit_others_posts'  => 'edit_others_posts',
+			'publish_posts'      => 'publish_posts',
+			'read_private_posts' => 'read_private_posts',
+		);
+		$order_args   = array(
+			'label'               => esc_html__( 'Plan', 'jetpack' ),
+			'description'         => esc_html__( 'Memberships plans', 'jetpack' ),
+			'supports'            => array( 'title', 'custom-fields', 'content' ),
+			'hierarchical'        => false,
+			'public'              => false,
+			'show_ui'             => false,
+			'show_in_menu'        => false,
+			'show_in_admin_bar'   => false,
+			'show_in_nav_menus'   => false,
+			'can_export'          => true,
+			'has_archive'         => false,
+			'exclude_from_search' => true,
+			'publicly_queryable'  => false,
+			'rewrite'             => false,
+			'capabilities'        => $capabilities,
+			'show_in_rest'        => false,
+		);
+		register_post_type( self::$post_type_plan, $order_args );
+	}
+
+	/**
+	 * Allows custom post types to be used by REST API.
+	 *
+	 * @param array $post_types - other post types.
+	 *
+	 * @see hook 'rest_api_allowed_post_types'
+	 * @return array
+	 */
+	public function allow_rest_api_types( $post_types ) {
+		$post_types[] = self::$post_type_plan;
+
+		return $post_types;
+	}
+
+	/**
+	 * Allows custom meta fields to sync.
+	 *
+	 * @param array $post_meta - previously changet post meta.
+	 *
+	 * @return array
+	 */
+	public function allow_sync_post_meta( $post_meta ) {
+		$meta_keys = array_map(
+			array( $this, 'return_meta' ),
+			$this->get_plan_property_mapping()
+		);
+		return array_merge( $post_meta, array_values( $meta_keys ) );
+	}
+
+	public function return_meta( $map ) {
+		return $map['meta'];
+	}
+	/**
+	 * Callback that parses the membership purchase shortcode.
+	 *
+	 * @param array       $attrs - attributes in the shortcode. `id` here is the CPT id of the plan.
+	 * @param string|bool $content - needed for the callback.
+	 *
+	 * @return string|void
+	 */
+	public function render_button( $attrs ) {
+		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name);
+
+		if ( empty( $attrs['planId'] ) ) {
+			return;
+		}
+		$id = $attrs['planId'];
+		$product = get_post( $id );
+		if ( ! $product || is_wp_error( $product ) ) {
+			return;
+		}
+		if ( $product->post_type !== self::$post_type_plan || 'trash' === $product->post_status ) {
```

Comment:
Does `trash ===` cover everything security wise? Shouldn't we check for `public !==`?

Reply:
There is no way of making this a draft or otherwise.
Not even security is our concern, since this serves as a local cache for BD data.
Manipulating this is proxied by access methods on wpcom.
I may need to have to rework this, but definitely there is no danger here.

Reply:
Is checking for `public !==` somehow bad idea, even if there are no security concerns?

Reply:
Thit has been tested multiple times and I am certain it works as intended.
I am afraid of touching it, since the post status is a very hacky thing - for example the `future` status seems like it can introduce problems in this scenario.
I would remove 'trash' check even, but this will be needed for omitting junk data created by inconsistent versions of this code.

Reply:
I was pointing out this just because I remember we had to fix some bugs around Simple payments exactly because of similar "trash" status check. While it probably worked fine initially, original author probably hadn't seen how someone could extend the code/API and leave easy way for regressions.

That said, this case might be different and it's up to you of course. 🙂

Reply:
(For a reference https://github.com/Automattic/wp-calypso/pull/28815/files#r235933564)

Reply:
Sorry, I did not follow this particular regression, but from what I understand from a quick scan:

- Data was stored in CPTs for shortcode
- Upon introducing a block, the initial solution attempted to store a copy of the same data in block attributes leading to source of truth problems, especially in block duplication
- This was resolved by introducing a "draft" status
- Which in turn, was clashing with that !== `trash` status check

Is there more to it? I'm having trouble finding a more comprehensive writeup.
I am all for avoiding regressions, but I don't want to introduce new ones while at it.

If there is an additional benefit of keeping the logic similar to simple payments without visible downsides, I'll implement it 👍